core/src/main/java/org/elasticsearch/cluster/metadata/IndexGraveyard.java

```diff
@@ -0,0 +1,449 @@
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.cluster.metadata;
+
+import org.elasticsearch.cluster.Diff;
+import org.elasticsearch.cluster.Diffable;
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamInputReader;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.xcontent.ObjectParser;
+import org.elasticsearch.common.xcontent.ToXContent;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.index.Index;
+
+import java.io.IOException;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+/**
+ * A collection of tombstones for explicitly marking indices as deleted in the cluster state.
+ *
+ * The cluster state contains a list of index tombstones for indices that have been
+ * deleted in the cluster.  Because cluster states are processed asynchronously by
+ * nodes and a node could be removed from the cluster for a period of time, the
+ * tombstones remain in the cluster state for a fixed period of time, after which
+ * they are purged.
+ */
+final public class IndexGraveyard implements ToXContent, Diffable<IndexGraveyard> {
+
+    // X-Content key for index tombstones
+    public static final String TOMBSTONES_KEY = "index_tombstones";
+
+    // Setting for the maximum tombstones allowed in the cluster state;
+    // prevents the cluster state size from exploding too large, but it opens the
+    // very unlikely risk that if there are greater than MAX_TOMBSTONES index
+    // deletions while a node was offline, when it comes back online, it will have
+    // missed index deletions that it may need to process.
+    private static final int MAX_TOMBSTONES = 10000;
+
+    // Setting for the expiration window of tombstone entries.  Any tombstone entries
+    // that have been in the list longer than EXPIRATION_WINDOW will be purged on the
+    // next cluster update.
+    static final long EXPIRATION_WINDOW = 1000 * 60 * 60 * 24 * 14; // 14 days
```

Comment:
For the default value, you can more safely use `TimeUnit.DAYS.convert(14, TimeUnit.MILLISECONDS)`.